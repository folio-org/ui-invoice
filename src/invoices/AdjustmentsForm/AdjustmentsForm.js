import React, { useState } from 'react';
import { FormattedMessage } from 'react-intl';
import {
  Field,
  FieldArray,
} from 'redux-form';
import PropTypes from 'prop-types';
import { find } from 'lodash';

import {
  Button,
  ButtonGroup,
  Card,
  Col,
  IconButton,
  KeyValue,
  RepeatableField,
  Row,
  Selection,
  TextField,
} from '@folio/stripes/components';
import {
  FieldSelect,
  FundDistributionFields,
  parseNumberFieldValue,
} from '@folio/stripes-acq-components';

import {
  ADJUSTMENT_PRORATE_OPTIONS,
  ADJUSTMENT_PRORATE_VALUES,
  ADJUSTMENT_RELATION_TO_TOTAL_OPTIONS,
  ADJUSTMENT_TYPE_VALUES,
} from '../../common/constants';
import {
  getAdjustmentPresetOptions,
  validateRequired,
} from '../../common/utils';

const getAdjustmentFromPreset = ({ description, prorate, relationToTotal, type, defaultAmount }) => ({
  description,
  prorate,
  relationToTotal,
  type,
  value: defaultAmount,
});

const AdjustmentsForm = ({ adjustmentsPresets, disabled, isLineAdjustments }) => {
  const [adjPreset, setAdjPreset] = useState();
  const onAdd = (fields) => {
    const newAdjustment = adjPreset
      ? getAdjustmentFromPreset(adjPreset.adjustment)
      : { type: ADJUSTMENT_TYPE_VALUES.amount };

    if (isLineAdjustments) delete newAdjustment.prorate;

    fields.push(newAdjustment);
  };

  // eslint-disable-next-line react/prop-types
  const renderTypeToggle = ({ input: { value, onChange } }) => {
    return (
      <KeyValue label={<FormattedMessage id="ui-invoice.settings.adjustments.type" />}>
        <ButtonGroup
          fullWidth
          data-test-adjustments-type
        >
          <Button
            onClick={() => onChange(ADJUSTMENT_TYPE_VALUES.percent)}
            buttonStyle={value === ADJUSTMENT_TYPE_VALUES.percent ? 'primary' : 'default'}
            data-test-adjustments-type-percent
            disabled={disabled}
          >
            <FormattedMessage id="ui-invoice.adjustment.type.sign.percent" />
          </Button>
          <Button
            onClick={() => onChange(ADJUSTMENT_TYPE_VALUES.amount)}
            buttonStyle={value === ADJUSTMENT_TYPE_VALUES.amount ? 'primary' : 'default'}
            data-test-adjustments-type-amount
            disabled={disabled}
          >
            <FormattedMessage id="ui-invoice.adjustment.type.sign.amount" />
          </Button>
        </ButtonGroup>
      </KeyValue>
    );
  };

  const renderAdjustment = (elem, index, fields) => {
    const onRemove = () => {
      fields.remove(index);
    };
    const adjustment = fields.get(index);
    const showFundDistribution = !isLineAdjustments
      && adjustment.prorate === ADJUSTMENT_PRORATE_VALUES.notProrated
      && adjustment.type === ADJUSTMENT_TYPE_VALUES.amount;

    return (
      <Card
        headerEnd={(
          <FormattedMessage id="stripes-components.deleteThisItem">
            {label => (
              <IconButton
                data-test-repeatable-field-remove-item-button
                icon="trash"
                onClick={onRemove}
                size="medium"
                disabled={disabled}
                ariaLabel={label}
              />
            )}
          </FormattedMessage>
        )}
        headerStart={(
          <FormattedMessage
            id="ui-invoice.adjustment.headerLabel"
            values={{ index: index + 1 }}
          />
        )}
      >
        <Row>
          <Col
            data-test-adjustment-description
            xs
          >
            <Field
              component={TextField}
              fullWidth
              label={<FormattedMessage id="ui-invoice.adjustment.description" />}
              name={`${elem}.description`}
              required
              validate={validateRequired}
              disabled={disabled}
            />
          </Col>
          <Col
            data-test-adjustment-amount
            xs
          >
            <Field
              component={TextField}
              fullWidth
              label={<FormattedMessage id="ui-invoice.adjustment.amount" />}
              name={`${elem}.value`}
              required
              type="number"
              parse={parseNumberFieldValue}
              validate={validateRequired}
              disabled={disabled}
            />
          </Col>
          <Col xs>
            <Field
              label="label"
              name={`${elem}.type`}
              component={renderTypeToggle}
            />
          </Col>
          {
            !isLineAdjustments && (
              <Col xs>
                <FieldSelect
                  label={<FormattedMessage id="ui-invoice.settings.adjustments.prorate" />}
                  name={`${elem}.prorate`}
                  dataOptions={ADJUSTMENT_PRORATE_OPTIONS}
                  required
                  validate={validateRequired}
                  disabled={disabled}
                />
              </Col>
            )
          }
          <Col
            data-test-adjustment-relation-to-total
            xs
          >
            <FieldSelect
              label={<FormattedMessage id="ui-invoice.settings.adjustments.relationToTotal" />}
              name={`${elem}.relationToTotal`}
              dataOptions={ADJUSTMENT_RELATION_TO_TOTAL_OPTIONS}
              required
              validate={validateRequired}
              disabled={disabled}
            />
          </Col>
        </Row>
        {showFundDistribution && (
          <FundDistributionFields
            disabled={disabled}
            fundDistribution={adjustment.fundDistributions}
            name={`${elem}.fundDistributions`}
            totalAmount={adjustment.value}
          />
        )}
      </Card>
    );
  };

  return (
    <Row>
      <Col xs={12}>
        <Selection
          dataOptions={getAdjustmentPresetOptions(adjustmentsPresets)}
          label={<FormattedMessage id="ui-invoice.adjustment.presetAdjustment" />}
          onChange={(id) => setAdjPreset(find(adjustmentsPresets, { id }))}
          disabled={disabled}
        />
      </Col>
      <Col xs={12}>
        <FieldArray
          addLabel={<FormattedMessage id="ui-invoice.button.addAdjustment" />}
          canAdd={!disabled}
          canRemove={!disabled}
          component={RepeatableField}
          id="adjustments"
          name="adjustments"
          onAdd={onAdd}
          onRemove={false}
          renderField={renderAdjustment}
        />
      </Col>
    </Row>
  );
};

AdjustmentsForm.propTypes = {
  adjustmentsPresets: PropTypes.arrayOf(PropTypes.object),
  disabled: PropTypes.bool,
  isLineAdjustments: PropTypes.bool,
};

AdjustmentsForm.defaultProps = {
  adjustmentsPresets: [],
  disabled: false,
  isLineAdjustments: false,
};

export default AdjustmentsForm;
