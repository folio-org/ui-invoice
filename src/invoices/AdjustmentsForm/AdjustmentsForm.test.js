import { MemoryRouter } from 'react-router-dom';

import { render, screen, waitFor } from '@folio/jest-config-stripes/testing-library/react';
import stripesFinalForm from '@folio/stripes/final-form';

import { adjustment } from '../../../test/jest/fixtures';

import AdjustmentsForm from './AdjustmentsForm';

jest.mock('../AdjustmentsDetails', () => jest.fn(() => 'AdjustmentsDetails'));

const TestForm = stripesFinalForm({})(
  (props) => {
    return (
      <form>
        <AdjustmentsForm {...props} />
      </form>
    );
  },
);

const renderForm = ({
  isNonInteractive = false,
  initialValues = {},
  change = jest.fn(),
  ...props
} = {}) => render(
  <TestForm
    onSubmit={jest.fn()}
    initialValues={initialValues}
    isNonInteractive={isNonInteractive}
    change={change}
    {...props}
  />,
  { wrapper: MemoryRouter },
);

describe('AdjustmentsForm', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  // TODO: release blocker: enable after release
  xit('should render correct structure with defined adjustments', async () => {
    const { asFragment } = renderForm({ initialValues: { adjustments: [adjustment] } });

    await screen.findByText('ui-invoice.adjustment.description');

    expect(asFragment()).toMatchSnapshot();
  });

  // TODO: release blocker: enable after release
  xit('should render correct structure without adjustments', async () => {
    const { asFragment } = renderForm();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display adjustments details in non interactive mode', async () => {
    renderForm({ isNonInteractive: true });

    expect(screen.getByText('AdjustmentsDetails')).toBeDefined();
  });

  it('should reset expenseClassId on fiscalYearId changed', async () => {
    const change = jest.fn();

    renderForm({
      isNonInteractive: true,
      initialValues: { adjustments: [adjustment] },
      fiscalYearId: 'fiscalYearId',
      adjustments: [{
        'type': 'Amount',
        'description': '111',
        'value': 1,
        'relationToTotal': 'In addition to',
        'prorate': 'Not prorated',
        'fundDistributions': [{
          'distributionType': 'percentage',
          'value': 100,
          'fundId': '7fbd5d84-62d1-44c6-9c45-6cb173998bbd',
          'code': 'AFRICAHIST',
          'encumbrance': null,
          'expenseClassId': '1bcc3247-99bf-4dca-9b0f-7bc51a2998c2',
        }],
      }],
      isFiscalYearChanged: true,
      change,
    });

    expect(screen.getByText('AdjustmentsDetails')).toBeDefined();
    await waitFor(() => expect(change).toHaveBeenCalledWith('adjustments[0].fundDistributions[0].expenseClassId', null));
  });
});
