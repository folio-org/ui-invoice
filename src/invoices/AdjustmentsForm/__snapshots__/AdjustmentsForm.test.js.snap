// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjustmentsForm should render correct structure with defined adjustments 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div>
          <label
            class="label"
            for="downshift-:r1:-input"
            id="sl-label-selection-:r0:"
          >
            ui-invoice.adjustment.presetAdjustment
          </label>
          <div
            class="selectionControlContainer"
          >
            <div
              class="inputGroup"
            >
              <button
                aria-controls="downshift-:r1:-menu"
                aria-expanded="false"
                aria-labelledby="sl-label-selection-:r0: selected-selection-:r0:-item"
                class="selectionControl formControl"
                id="selection-:r0:"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  stripes-components.selection.controlLabel
                </span>
                <div
                  class="singleValue"
                  id="selected-selection-:r0:-item"
                />
              </button>
              <div
                class="selectionEndControls"
              >
                <div
                  class="textFieldIcon"
                >
                  <span>
                    Icon
                  </span>
                </div>
              </div>
            </div>
            <div
              role="alert"
            />
          </div>
          <div
            class="overlay"
            hidden=""
          >
            <div
              class="selectionListRoot"
              id="sl-container-selection-:r0:"
              style="width: 0px;"
            >
              <div
                class="selectionFilterContainer"
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-controls="downshift-:r1:-menu"
                  aria-expanded="false"
                  aria-label="stripes-components.selection.filterOptionsLabel"
                  aria-labelledby="downshift-:r1:-label"
                  autocomplete="off"
                  class="selectionFilter"
                  id="downshift-:r1:-input"
                  placeholder="stripes-components.selection.filterOptionsPlaceholder"
                  role="combobox"
                  type="text"
                  value=""
                />
              </div>
              <ul
                aria-labelledby="sl-label-selection-:r0:"
                class="selectionList"
                id="downshift-:r1:-menu"
                role="listbox"
                style="max-height: 174px;"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xs-12"
      >
        <fieldset
          class="repeatableField hasMargin"
          data-test-repeatable-field="true"
          id="adjustments"
          tabindex="-1"
        >
          <ul
            class="repeatableFieldList"
            data-test-repeatable-field-list="true"
          >
            <li
              class="repeatableFieldItem"
              data-test-repeatable-field-list-item="true"
            >
              <div
                class="repeatableFieldItemContent"
              >
                <div
                  class="card default"
                >
                  <div
                    class="header default"
                    data-test-card-header="true"
                  >
                    <span
                      class="headerStart"
                      data-test-card-header-start="true"
                    >
                      ui-invoice.adjustment.headerLabel
                    </span>
                    <span
                      data-test-card-header-end="true"
                    >
                      <button
                        aria-label="stripes-components.deleteThisItem"
                        class="iconButton medium"
                        data-test-repeatable-field-remove-item-button="true"
                        icon="trash"
                        type="button"
                      >
                        <span
                          class="iconButtonInner mediumOffset"
                          data-test-repeatable-field-remove-item-button="true"
                        >
                          <span>
                            Icon
                          </span>
                        </span>
                      </button>
                    </span>
                  </div>
                  <div
                    class="body "
                    data-test-card-body="true"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col-xs"
                        data-test-adjustment-description="true"
                      >
                        <div
                          class="textField"
                        >
                          <label
                            class="label"
                            for="adjustments[0].description"
                          >
                            ui-invoice.adjustment.description
                            <span
                              aria-hidden="true"
                              class="asterisk"
                            >
                              *
                            </span>
                          </label>
                          <div
                            class="formControl inputGroup"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              autocapitalize="off"
                              autocomplete="off"
                              autocorrect="off"
                              error=""
                              id="adjustments[0].description"
                              name="adjustments[0].description"
                              required=""
                              spellcheck="false"
                              type="text"
                              value=""
                              warning=""
                            />
                          </div>
                          <div
                            role="alert"
                          />
                        </div>
                      </div>
                      <div
                        class="col-xs"
                        data-test-adjustment-amount="true"
                      >
                        <div
                          class="textField"
                        >
                          <label
                            class="label"
                            for="adjustments[0].value"
                          >
                            ui-invoice.adjustment.value
                            <span
                              aria-hidden="true"
                              class="asterisk"
                            >
                              *
                            </span>
                          </label>
                          <div
                            class="formControl inputGroup"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              autocapitalize="off"
                              autocomplete="off"
                              autocorrect="off"
                              error=""
                              id="adjustments[0].value"
                              name="adjustments[0].value"
                              required=""
                              spellcheck="false"
                              type="number"
                              value="25.5"
                              warning=""
                            />
                          </div>
                          <div
                            role="alert"
                          />
                        </div>
                      </div>
                      <div
                        class="col-xs"
                      >
                        <div
                          class="kvRoot"
                        >
                          <div
                            class="kvLabel"
                          >
                            ui-invoice.settings.adjustments.type
                          </div>
                          <div
                            class="kvValue"
                            data-test-kv-value="true"
                          >
                            <div
                              class="buttonGroup fullWidth"
                              data-test-adjustments-type="true"
                            >
                              <button
                                class="button default"
                                data-test-adjustments-type-percent="true"
                                type="button"
                              >
                                <span
                                  class="inner"
                                >
                                  ui-invoice.adjustment.type.sign.percent
                                </span>
                              </button>
                              <button
                                class="button primary"
                                data-test-adjustments-type-amount="true"
                                type="button"
                              >
                                <span
                                  class="inner"
                                >
                                  $
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="col-xs"
                      >
                        <div
                          class="select inputGroup fullWidth"
                        >
                          <label
                            class="label"
                            for="select-1"
                            id="select-1-label"
                          >
                            ui-invoice.settings.adjustments.prorate
                            <span
                              aria-hidden="true"
                              class="asterisk"
                            >
                              *
                            </span>
                          </label>
                          <div
                            class="selectWrap"
                          >
                            <select
                              aria-invalid="false"
                              aria-labelledby="select-1-label"
                              aria-required="true"
                              class="selectControl formControl"
                              id="select-1"
                              name="adjustments[0].prorate"
                              required=""
                            >
                              <option
                                value=""
                              >
                                stripes-acq-components.label.emptyValue
                              </option>
                              <option
                                value="By line"
                              >
                                ui-invoice.adjustments.prorate.byLine
                              </option>
                              <option
                                value="By amount"
                              >
                                ui-invoice.adjustments.prorate.byAmount
                              </option>
                              <option
                                value="By quantity"
                              >
                                ui-invoice.adjustments.prorate.byQuantity
                              </option>
                              <option
                                value="Not prorated"
                              >
                                ui-invoice.adjustments.prorate.notProrated
                              </option>
                            </select>
                            <span>
                              Icon
                            </span>
                          </div>
                          <div
                            role="alert"
                          />
                        </div>
                      </div>
                      <div
                        class="col-xs"
                        data-test-adjustment-relation-to-total="true"
                      >
                        <div
                          class="select inputGroup fullWidth"
                        >
                          <label
                            class="label"
                            for="select-2"
                            id="select-2-label"
                          >
                            ui-invoice.settings.adjustments.relationToTotal
                            <span
                              aria-hidden="true"
                              class="asterisk"
                            >
                              *
                            </span>
                          </label>
                          <div
                            class="selectWrap"
                          >
                            <select
                              aria-invalid="false"
                              aria-labelledby="select-2-label"
                              aria-required="true"
                              class="selectControl formControl"
                              id="select-2"
                              name="adjustments[0].relationToTotal"
                              required=""
                            >
                              <option
                                value=""
                              >
                                stripes-acq-components.label.emptyValue
                              </option>
                              <option
                                value="In addition to"
                              >
                                ui-invoice.adjustments.relationToTotal.inAdditionTo
                              </option>
                              <option
                                value="Included in"
                              >
                                ui-invoice.adjustments.relationToTotal.includedIn
                              </option>
                              <option
                                value="Separate from"
                              >
                                ui-invoice.adjustments.relationToTotal.separateFrom
                              </option>
                            </select>
                            <span>
                              Icon
                            </span>
                          </div>
                          <div
                            role="alert"
                          />
                        </div>
                      </div>
                      <div
                        class="col-xs"
                        data-test-adjustment-export-to-accounting="true"
                      >
                        <div
                          class="checkbox inline vertical"
                          data-test-checkbox="true"
                        >
                          <label
                            class="label label checkboxInteractionStylesControl"
                            for="checkbox-3"
                          >
                            <span
                              class="labelText"
                            >
                              ui-invoice.settings.adjustments.exportToAccounting
                            </span>
                            <span
                              class="inner checkboxInteractionStyles"
                            >
                              <input
                                aria-invalid="false"
                                class="input"
                                id="checkbox-3"
                                name="adjustments[0].exportToAccounting"
                                type="checkbox"
                                value=""
                              />
                              <span
                                class="checkboxIcon"
                              >
                                <svg
                                  height="10"
                                  viewBox="0 0 14 14"
                                  width="10"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M10.395 4.617L5.63 9.383 3.605 7.358"
                                    fill="none"
                                    stroke-linecap="round"
                                    stroke-linejoin="round"
                                    stroke-miterlimit="10"
                                    stroke-width="1.5"
                                  />
                                </svg>
                              </span>
                            </span>
                          </label>
                          <div
                            role="alert"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </li>
          </ul>
          <button
            class="button default"
            data-test-repeatable-field-add-item-button="true"
            id="adjustments-add-button"
            type="button"
          >
            <span
              class="inner"
            >
              ui-invoice.button.addAdjustment
            </span>
          </button>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`AdjustmentsForm should render correct structure without adjustments 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div>
          <label
            class="label"
            for="downshift-:r3:-input"
            id="sl-label-selection-:r2:"
          >
            ui-invoice.adjustment.presetAdjustment
          </label>
          <div
            class="selectionControlContainer"
          >
            <div
              class="inputGroup"
            >
              <button
                aria-controls="downshift-:r3:-menu"
                aria-expanded="false"
                aria-labelledby="sl-label-selection-:r2: selected-selection-:r2:-item"
                class="selectionControl formControl"
                id="selection-:r2:"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  stripes-components.selection.controlLabel
                </span>
                <div
                  class="singleValue"
                  id="selected-selection-:r2:-item"
                />
              </button>
              <div
                class="selectionEndControls"
              >
                <div
                  class="textFieldIcon"
                >
                  <span>
                    Icon
                  </span>
                </div>
              </div>
            </div>
            <div
              role="alert"
            />
          </div>
          <div
            class="overlay"
            hidden=""
          >
            <div
              class="selectionListRoot"
              id="sl-container-selection-:r2:"
            >
              <div
                class="selectionFilterContainer"
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-controls="downshift-:r3:-menu"
                  aria-expanded="false"
                  aria-label="stripes-components.selection.filterOptionsLabel"
                  aria-labelledby="downshift-:r3:-label"
                  autocomplete="off"
                  class="selectionFilter"
                  id="downshift-:r3:-input"
                  placeholder="stripes-components.selection.filterOptionsPlaceholder"
                  role="combobox"
                  type="text"
                  value=""
                />
              </div>
              <ul
                aria-labelledby="sl-label-selection-:r2:"
                class="selectionList"
                id="downshift-:r3:-menu"
                role="listbox"
                style="max-height: 174px;"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xs-12"
      >
        <fieldset
          class="repeatableField hasMargin"
          data-test-repeatable-field="true"
          id="adjustments"
          tabindex="-1"
        >
          <button
            class="button default"
            data-test-repeatable-field-add-item-button="true"
            id="adjustments-add-button"
            type="button"
          >
            <span
              class="inner"
            >
              ui-invoice.button.addAdjustment
            </span>
          </button>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
